### DISCUSSION TITLE
How to debug the ColossalAI script in PyCharm

### DISCUSSION THREAD

**Discussion by songyuc** [hpcaitech/ColossalAI#1055] [Community | Q&A]
Hi, guys,
I am a learner of ColossalAI, and I find the ColossalAI script needs to run under the "colossalai run" program as:
```bash
# with engine
colossalai run --nproc_per_node 1 train.py
```
So, how can I debug it in PyCharm, which require a python interpreter to run a script.

Your answer and guide will be appreciated!

**Comment by YuliangLiu0306** [comment]
If you want to debug your model, I suggest you use single gpu mode, then you can use "python XXX.py" to run your program. If you have to debug program with multiple devices, you can use debugpy in vscode, or something have same effect in pycharm.

**Reply by songyuc** [reply]
So, if I use PyCharm, do I need to add other settings? Commonly, `.py` script needs to run as `python train.py`. I will learn to use ColossalAI on my single 2080Ti.

**Reply by FrankLeeeee** [reply]
`colossalai run` will call `python train.py` internally.

**Reply by songyuc** [reply]
> `colossalai run` will call `python train.py` internally.

But `collossalai run` might not be able to trigger the debug of PyCharm, which regards `python` as a default interpreter.

**Reply by YuliangLiu0306** [reply]
> So, if I use PyCharm, do I need to add other settings? Commonly, `.py` script needs to run as `python train.py`. I will learn to use ColossalAI on my single 2080Ti.

You don't have to add any setting with single device.

**Reply by songyuc** [reply]
> > So, if I use PyCharm, do I need to add other settings? Commonly, `.py` script needs to run as `python train.py`. I will learn to use ColossalAI on my single 2080Ti.
> 
> You don't have to add any setting with single device.

Thank you sincerely!

**Reply by songyuc** [reply]
Hi, @YuliangLiu0306 , I met a problem today when running with `python train.py`, as below,
```bash 
/home/user/software/python/anaconda/anaconda3/envs/conda-general/bin/python /home/user/***/***
/ColossalAI-Examples/image/resnet/train.py
Traceback (most recent call last):
  File "/home/user/software/python/anaconda/anaconda3/envs/conda-general/lib/python3.10/site-packages/colossalai/initialize.py", line 210, in launch_from_torch
    rank = int(os.environ['RANK'])
  File "/home/user/software/python/anaconda/anaconda3/envs/conda-general/lib/python3.10/os.py", line 679, in __getitem__
    raise KeyError(key) from None
KeyError: 'RANK'

During handling of the above exception, another exception occurred:

...

RuntimeError: Could not find 'RANK' in the torch environment, visit https://www.colossalai.org/ for more information on launching with torch
```
Is this error due to the absence of environment variable `RANK` in my Ubuntu?

**Reply by binmakeswell** [reply]
Reply in #1074

**Reply by tongping** [reply]
> If you want to debug your model, I suggest you use single gpu mode, then you can use "python XXX.py" to run your program. If you have to debug program with multiple devices, you can use debugpy in vscode, or something have same effect in pycharm.

As suggested, I was able to debug the model when using single gpu mode. However, sometimes we cannot use single gpu mode, such as tensor parallelism. How to debug in such situations?

**Reply by tongping** [reply]
> Hi, @YuliangLiu0306 , I met a problem today when running with `python train.py`, as below,
> 
> ```shell
> /home/user/software/python/anaconda/anaconda3/envs/conda-general/bin/python /home/user/***/***
> /ColossalAI-Examples/image/resnet/train.py
> Traceback (most recent call last):
>   File "/home/user/software/python/anaconda/anaconda3/envs/conda-general/lib/python3.10/site-packages/colossalai/initialize.py", line 210, in launch_from_torch
>     rank = int(os.environ['RANK'])
>   File "/home/user/software/python/anaconda/anaconda3/envs/conda-general/lib/python3.10/os.py", line 679, in __getitem__
>     raise KeyError(key) from None
> KeyError: 'RANK'
> 
> During handling of the above exception, another exception occurred:
> 
> ...
> 
> RuntimeError: Could not find 'RANK' in the torch environment, visit https://www.colossalai.org/ for more information on launching with torch
> ```
> 
> Is this error due to the absence of environment variable `RANK` in my Ubuntu?
Yes, that is the reason. 
This is easy to fix. You could define the environment variables in your code like this: 
   os.environ['RANK'] = '0'
    os.environ['LOCAL_RANK'] = '0'
    os.environ['WORLD_SIZE'] = '2'
    os.environ['MASTER_ADDR'] = '1d2d8b0466e2'
    os.environ['MASTER_PORT'] = '42043'

**Reply by kangxi1818** [reply]
what does '1d2d8b0466e2' and '42043' means and how to get that?
